java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/BinaryRelation.java

```diff
@@ -21,336 +21,312 @@
     }
-    return RelationState.UNDETERMINED;
+    if (b.kind == LESS_THAN) {
+      return lessThanTransitiveBuilder(b, a);
+    }
+    if (a.kind == GREATER_THAN_OR_EQUAL && b.kind == GREATER_THAN_OR_EQUAL) {
+      BinaryRelation greaterThanEqual = greaterThanEqualTransitiveBuilder(a, b);
+      return greaterThanEqual != null ? greaterThanEqual : greaterThanEqualTransitiveBuilder(b, a);
+    }
+    return null;
   }
 
-  protected abstract RelationState isImpliedBy(BinaryRelation relation);
-
-  protected abstract RelationState impliesEqual();
-
-  protected abstract RelationState impliesNotEqual();
-
-  protected abstract RelationState impliesMethodEquals();
-
-  protected abstract RelationState impliesNotMethodEquals();
-
-  protected abstract RelationState impliesGreaterThan();
-
-  protected abstract RelationState impliesGreaterThanOrEqual();
-
-  protected abstract RelationState impliesLessThan();
-
-  protected abstract RelationState impliesLessThanOrEqual();
+  private static boolean isEqualityRelation(Kind kind) {
+    return kind == EQUAL || kind == METHOD_EQUALS;
+  }
 
-  /**
-   * Returns a new relation resulting of the transitive combination of the receiver with the supplied relation.
-   * Schema aRb & bSc => aTc
-   * where aRb is the relation described by the receiver, bSc that of the supplied relation, aTc is the returned relation.
-   * 
-   * @param relation another relation whose left operand is the same as the receiver's right operand
-   * @return the resulting relation or null if the conjunction of both relations does not bring any new information
-   */
-  @CheckForNull
-  protected abstract BinaryRelation combinedAfter(BinaryRelation relation);
+  private static BinaryRelation equalityTransitiveBuilder(BinaryRelation a, BinaryRelation b) {
+    BinaryRelation equality = isEqualityRelation(a.kind) ? a : b;
+    BinaryRelation other = equality == a ? b : a;
+    SymbolicValue transitiveOperand = equality.differentOperand(other);
+    boolean leftTransitive = equality.hasOperand(other.leftOp);
 
-  @CheckForNull
-  protected abstract BinaryRelation combinedWithEqual(EqualRelation relation);
+    if (equality.kind == METHOD_EQUALS && other.kind == EQUAL) {
+      return equalityTransitiveBuilder(other, equality);
+    }
 
-  @CheckForNull
-  protected abstract BinaryRelation combinedWithNotEqual(NotEqualRelation relation);
+    BinaryRelation deduced = binaryRelation(other.kind,
+      leftTransitive ? transitiveOperand : other.leftOp,
+      leftTransitive ? other.rightOp : transitiveOperand);
 
-  @CheckForNull
-  protected abstract BinaryRelation combinedWithMethodEquals(MethodEqualsRelation relation);
-
-  @CheckForNull
-  protected abstract BinaryRelation combinedWithNotMethodEquals(NotMethodEqualsRelation relation);
+    // FIXME remove this sanity check
+    Preconditions.checkState(!a.equals(deduced), "Deduced relation " + deduced + " is equal to input relation " + a);
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Use the built-in formatting to construct this argument. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2629)